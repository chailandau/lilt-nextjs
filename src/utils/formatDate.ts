/**
 * Formats a given timestamp into a string representation of the date.
 *
 * @param timestamp - The timestamp to be formatted.
 * @returns The formatted date string.
 */
export const formatDate = (timestamp: string | number | Date) => {
    const date = new Date(timestamp);

    return date.toLocaleDateString('en-US', {
        year: 'numeric',
        month: 'long',
        day: '2-digit'
    });
};
