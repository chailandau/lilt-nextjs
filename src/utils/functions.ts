/**
 * Determine if provided image source URL is an SVG.
 *
 * @param {string} sourceUrl - The URL of the image to check.
 * @return {boolean} True if the URL ends with ".svg", false otherwise.
 */
export const isSvg = (sourceUrl: string): boolean => sourceUrl.endsWith('.svg');

