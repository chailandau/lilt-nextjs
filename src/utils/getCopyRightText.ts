/**
 * Replaces all occurrences of "%CURRENT_YEAR%" in the given string with the current year. If the string is empty, returns a default copyright text.
 *
 * @param str - The string to be modified.
 * @returns The modified string with all occurrences of "%CURRENT_YEAR%" replaced with the current year.
 */

const isLimg = process.env.NEXT_PUBLIC_THEME === 'green-theme';

export const getCopyrightText = (str?: string | null): string => {
    if (!str) {
        str = `Copyright Â© %CURRENT_YEAR% ${
            isLimg ? 'Long Island Mini Golf' : 'Long Island Laser Tag'
        }. All Rights Reserved.`;
    }
    const currentYear = new Date().getFullYear().toString();

    return str.replaceAll('%CURRENT_YEAR%', currentYear);
};
