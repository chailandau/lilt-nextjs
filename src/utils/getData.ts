import { graphQLClient } from '@/api/graphqlClient';
import { Query } from '@/api/graphqlTypes';

/**
 * Retrieves data from the API using the provided GraphQL query.
 *
 * @param QUERY - The GraphQL query to be executed.
 * @param slug - (Optional) The slug to be used as a variable in the query.
 * @returns A promise that resolves to the data retrieved from the API.
 */
export const getData = async (QUERY: string, slug: string | null = null) => {
    try {
        const variables = slug !== null ? { slug } : {};

        return await graphQLClient.request<Query>(QUERY, variables);
    } catch (error) {
        console.error('GraphQL request error:', error);
        throw error;
    }
};
