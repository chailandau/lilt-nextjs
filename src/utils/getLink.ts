import {
    CallToAction,
    CallToAction_LinkType,
    Header_MenuItems_LinkType
} from '@/api/graphqlTypes';

interface GetLink {
    linkType?: CallToAction_LinkType | Header_MenuItems_LinkType | null;
    externalLink?: CallToAction['externalLink'];
    internalLink?: CallToAction['internalLink'];
}

/**
 * Returns appropriate link based on specified link type.
 *
 * @param linkType - Type of link to return.
 * @param externalLink - External link (if it exists).
 * @param internalLink - Internal link object (if it exists).
 * @return Appropriate link based on link type. Returns empty string if no link is found.
 */
export const getLink = ({ linkType, externalLink, internalLink }: GetLink) => {
    const isExternal =
        linkType === 'external' &&
        !externalLink?.includes('longislandlasertag.com');

    const isHome = linkType === 'internal' && internalLink?.slug === 'home';

    if (isExternal && externalLink) {
        return `${externalLink}`;
    } else if (internalLink?.slug) {
        return isHome
            ? `${process.env.NEXT_PUBLIC_BASE_URL}`
            : `${process.env.NEXT_PUBLIC_BASE_URL}/${internalLink.slug}`;
    } else {
        return '';
    }
};
