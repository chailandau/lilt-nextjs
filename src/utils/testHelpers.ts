import { fireEvent, render } from '@testing-library/react';
import { axe } from 'jest-axe';
import { ReactElement, cloneElement } from 'react';

type Component = ReactElement;

interface TestComponent {
    component: Component;
}

interface TestRole extends TestComponent {
    role: string;
}

interface TestTag extends TestComponent {
    htmlTag: string;
}

interface TestPropOptions extends TestTag {
    propName: string;
    propOptions: readonly string[];
}

interface TestTags extends TestComponent {
    tags: readonly string[];
}

interface TestRenderText extends TestRole {
    text: string;
}

/**
 * Maps through a list of prop options from a provided array, testing that each rendered component contains the specified prop option.
 *
 * @param component - Component to test.
 * @param propName - Name of prop to pass options to.
 * @param propOptions - Array of prop options to map through.
 * @param htmlTag - HTML tag of the component.
 */
export const testPropOptions = ({
    component,
    propName,
    propOptions,
    htmlTag
}: TestPropOptions): void => {
    propOptions.forEach((propOption) => {
        it(`has '${propOption}' class`, () => {
            const { container } = render(
                cloneElement(component, { [propName]: propOption })
            );
            expect(container.querySelector(htmlTag)).toHaveClass(propOption);
        });
    });
};

/**
 * Tests that `className` is applied to component if it exists.
 *
 * @param component - Component to test.
 * @param htmlTag - HTML tag of the component.
 */
export const testAppliesClass = ({ component, htmlTag }: TestTag): void => {
    it('applies class if it exists', () => {
        const { container } = render(
            cloneElement(component, { className: 'test' })
        );
        expect(container.querySelector(htmlTag)).toHaveClass('test');
    });
};

/**
 * Maps through a list of tags from a provided array, testing that each rendered component
 * contains the specified tag.
 *
 * @param component - Component to render.
 * @param tags - Array of tags to map through.
 */
export const testTags = ({ component, tags }: TestTags): void => {
    tags.forEach((tag) => {
        it(`has '${tag}' tag`, () => {
            const { container } = render(cloneElement(component, { as: tag }));
            expect(container.querySelector(tag)).toBeDefined();
        });
    });
};

/**
 * Tests that component renders with correct text.
 *
 * @param component - Component to render.
 * @param text - Text to render.
 */
export const testRenderText = ({ component, text }: TestRenderText): void => {
    it('renders with correct text', () => {
        const { container } = render(component);
        expect(container.textContent).toBe(text);
    });
};

/**
 * Tests that children render correctly inside a Higher Order Component (HOC) component.
 *
 * @param component - Component to test.
 */
export const testRenderChildren = ({ component }: TestComponent): void => {
    it('renders children inside HOC component', () => {
        const { getByText } = render(component);
        expect(getByText('Test')).toBeDefined();
    });
};

/**
 * Tests that keydown event is fired when key is pressed.
 *
 * @param component - Component to test.
 * @param role - Component's ARIA role.
 */
export const testKeyDown = ({ component, role }: TestRole): void => {
    it('should call onKeyDown when a key is pressed', () => {
        const onKeyDownMock = jest.fn();

        const { getByRole } = render(
            cloneElement(component, {
                onKeyDown: onKeyDownMock
            })
        );
        const element = getByRole(role);
        fireEvent.keyDown(element, { key: 'Enter', code: 13 });
        expect(onKeyDownMock).toHaveBeenCalledTimes(1);

        const eventObject = onKeyDownMock.mock.calls[0][0];
        expect(eventObject).toBeDefined();
        expect(eventObject.key).toBe('Enter');
    });
};

/**
 * Tests that component matches snapshot and does not have axe violations.
 *
 * @param component - Component to test.
 */
export const testAxeAndSnapshot = ({ component }: TestComponent) => {
    it('matches snapshot', () => {
        const { asFragment } = render(component);
        expect(asFragment()).toMatchSnapshot();
    });
    it('does not have axe violations', async () => {
        const { container } = render(component);
        expect(await axe(container)).toHaveNoViolations();
    });
};
