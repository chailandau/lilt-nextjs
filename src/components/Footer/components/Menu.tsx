import { FC } from 'react';

import { FooterProps } from '../Footer';

import Link from '@/atoms/Link';
import { getLink } from '@/utils/getLink';

const Menu: FC<FooterProps> = ({ menuItems }) => {
    const menuContent = menuItems?.map((menuItem) => {
        if (
            !menuItem.linkType ||
            (menuItem.linkType === 'external' && !menuItem.label) ||
            (menuItem.linkType === 'internal' &&
                (!menuItem?.internalLink?.slug ||
                    !menuItem.internalLink?.title))
        ) {
            return null;
        }

        const menuLink = getLink({
            linkType: menuItem?.linkType,
            externalLink: menuItem?.externalLink,
            internalLink: menuItem?.internalLink
        });

        const menuLabel =
            menuItem?.linkType === 'external'
                ? menuItem?.label
                : menuItem?.internalLink?.title;

        return (
            <li key={menuItem?.id}>
                <Link key={menuItem?.id} href={menuLink}>
                    {menuLabel}
                </Link>
            </li>
        );
    });

    return <ul>{menuContent}</ul>;
};

export default Menu;
