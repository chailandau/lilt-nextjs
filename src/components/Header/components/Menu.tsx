import { FC, KeyboardEvent } from 'react';

import { HeaderProps } from '../Header';

import styles from './Menu.module.scss';
import Submenu from './Submenu';

import Link from '@/atoms/Link';
import useStore from '@/store/useStore';
import { getLink } from '@/utils/getLink';

export interface MenuProps extends HeaderProps {
    /** Optional classname */
    className?: string;
}

const Menu: FC<MenuProps> = ({ menuItems, className = styles['menu'] }) => {
    const menuContent = menuItems?.map((menuItem, index) => {
        if (
            !menuItem.linkType ||
            (menuItem.linkType === 'external' && !menuItem.label) ||
            (menuItem.linkType === 'internal' &&
                (!menuItem?.internalLink?.slug ||
                    !menuItem.internalLink?.title))
        ) {
            return null;
        }
        const { setOpenSubmenu } = useStore();

        const handleKeyDown = (
            e: KeyboardEvent<HTMLAnchorElement | HTMLButtonElement>
        ) => {
            const isLastItem = index === menuItems?.length - 1;
            const isSubmenu = menuItem?.submenuItems === undefined;
            if (e.key === 'Tab' && !e.shiftKey && isLastItem && isSubmenu) {
                setOpenSubmenu(null);
            }
        };

        const menuLink = getLink({
            linkType: menuItem?.linkType,
            externalLink: menuItem?.externalLink,
            internalLink: menuItem?.internalLink
        });

        const overridePageName =
            menuItem?.overridePageName && menuItem?.internalCustomLabel;

        const internalMenuLabel = overridePageName
            ? menuItem?.internalCustomLabel
            : menuItem?.internalLink?.title;

        const menuLabel =
            menuItem?.linkType === 'external'
                ? menuItem?.label
                : internalMenuLabel;

        if (menuItem?.linkType === 'submenu') {
            return (
                menuItem?.label &&
                menuItem?.submenuItems &&
                menuItem.submenuItems.length > 0 && (
                    <Submenu key={menuItem?.id} menuItem={menuItem} />
                )
            );
        } else {
            return (
                <li key={menuItem?.id}>
                    <Link
                        key={menuItem?.id}
                        href={menuLink || ''}
                        className={styles['menu-link']}
                        onKeyDown={(e) => handleKeyDown(e)}
                    >
                        {menuLabel}
                    </Link>
                </li>
            );
        }
    });

    return <ul className={className}>{menuContent}</ul>;
};

export default Menu;
