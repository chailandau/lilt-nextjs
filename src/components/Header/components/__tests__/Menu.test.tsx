import { fireEvent, render, screen } from '@testing-library/react';

import Menu from '../Menu';

import {
    menuItemNoLabel,
    menuItemNoLinktype,
    menuItems,
    menuItemsInternalAndExternal
} from '@/__mocks__/Header.mock';
import { testAxeAndSnapshot } from '@/utils/testHelpers';

describe('Menu', () => {
    it('renders menu items correctly', () => {
        render(<Menu menuItems={menuItemsInternalAndExternal} />);

        const menuItemsElements = screen.getAllByRole('listitem');

        expect(menuItemsElements).toHaveLength(2);
    });

    it("doesn't render menu item if linkType isn't set", () => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        render(<Menu menuItems={menuItemNoLinktype as any} />);

        const menuItemsElements = screen.getByRole('list').querySelector('li');

        expect(menuItemsElements).toBeNull();
    });

    it("doesn't render menu item if linkType is external + label isn't set", () => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        render(<Menu menuItems={menuItemNoLabel} />);

        const menuItemsElements = screen.getByRole('list').querySelector('li');

        expect(menuItemsElements).toBeNull();
    });

    it('exits submenu on tab out', () => {
        render(<Menu menuItems={menuItems} />);

        const aboutMenu = screen.getByRole('button', {
            name: 'About'
        });

        fireEvent.keyDown(aboutMenu, { key: 'Enter' });

        const aboutSubmenu = aboutMenu.querySelector('ul');

        const ourTechLink = screen.getByRole('link', {
            name: 'Our Tech'
        });

        fireEvent.keyDown(ourTechLink, { key: 'Tab', shiftKey: false });

        expect(aboutSubmenu).not.toBeInTheDocument();
    });

    testAxeAndSnapshot({ component: <Menu menuItems={menuItems} /> });
});
