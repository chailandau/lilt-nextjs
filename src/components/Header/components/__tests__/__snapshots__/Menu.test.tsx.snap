// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="menu"
  >
    <li>
      <a
        class="link underline-hover menu-link"
        href="/booking"
        rel=""
        target=""
      >
        Booking
      </a>
    </li>
    <li>
      <a
        class="link underline-hover menu-link"
        href="/events"
        rel=""
        target=""
      >
        Events
      </a>
    </li>
    <li>
      <a
        class="link underline-hover menu-link"
        href="/schools-camps"
        rel=""
        target=""
      >
        Schools & Camps
      </a>
    </li>
    <li>
      <button
        class="button link submenu-btn"
      >
        About
        <svg
          class="gray sm"
        >
          <use
            href="/icons/icon-sprites.svg#caret"
          />
        </svg>
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
